pkg/virtualgarden/kube_api_server_certificates.go

```diff
@@ -0,0 +1,258 @@
+	if err != nil {
+		return err
+	}
+
+	_, err = o.deployKubeApiServerClientAdminCertificate(ctx, apiServerCACertificate, loadbalancer, checksums)
+	if err != nil {
+		return err
+	}
+
+	_, err = o.deployKubeApiServerMetricsScraperCertificate(ctx, apiServerCACertificate, checksums)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *operation) deleteKubeAPIServerCertificates(ctx context.Context) error {
+	o.log.Infof("Deleting secrets containing kube-apiserver certificates")
+
+	for _, name := range []string{
+		KubeApiServerSecretNameAggregatorCACertificate,
+		KubeApiServerSecretNameAggregatorClientCertificate,
+		KubeApiServerSecretNameApiServerCACertificate,
+		KubeApiServerSecretNameApiServerServerCertificate,
+		KubeApiServerSecretNameKubeControllerManagerCertificate,
+		KubeApiServerSecretNameClientAdminCertificate,
+		KubeApiServerSecretNameMetricsScraperCertificate,
+	} {
+		secret := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: o.namespace}}
+		if err := o.client.Delete(ctx, secret); client.IgnoreNotFound(err) != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (o *operation) deployKubeApiServerApiServerCACertificate(ctx context.Context, checksums map[string]string) (*secretsutil.Certificate, error) {
+	certConfig := &secretsutil.CertificateSecretConfig{
+		Name:       KubeApiServerSecretNameApiServerCACertificate,
+		CertType:   secretsutil.CACert,
+		CommonName: Prefix + ":ca:kube-apiserver",
+	}
+
+	cert, checksum, _, err := o.deployCertificate(ctx, certConfig, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	o.exports.KubeApiserverCaPem = string(cert.CertificatePEM)
+
+	checksums[ChecksumKeyKubeAPIServerCA] = checksum
+	return cert, err
+}
+
+func (o *operation) deployKubeApiServerApiServerServerCertificate(ctx context.Context, caCertificate *secretsutil.Certificate,
+	loadbalancer string, checksums map[string]string) (*secretsutil.Certificate, error) {
+	dnsAccessDomain := o.imports.VirtualGarden.KubeAPIServer.DnsAccessDomain
+
+	certConfig := &secretsutil.CertificateSecretConfig{
+		Name:       KubeApiServerSecretNameApiServerServerCertificate,
+		CertType:   secretsutil.ServerCert,
+		SigningCA:  caCertificate,
+		CommonName: Prefix + ":server:kube-apiserver",
+		DNSNames: []string{
+			"localhost",
+			KubeAPIServerServiceName,
+			KubeAPIServerServiceName + ".garden",
+			KubeAPIServerServiceName + ".garden.svc",
+			KubeAPIServerServiceName + ".garden.svc.cluster",
+			KubeAPIServerServiceName + ".garden.svc.cluster.local",
+			"kubernetes",
+			"kubernetes.default",
+			"kubernetes.default.svc",
+			"kubernetes.default.svc.cluster",
+			"kubernetes.default.svc.cluster.local",
+			fmt.Sprintf("api.%s", dnsAccessDomain),
+			fmt.Sprintf("gardener.%s", dnsAccessDomain),
+		},
+		IPAddresses: []net.IP{
+			net.ParseIP("127.0.0.1"),
+			net.ParseIP("100.64.0.1"),
+			net.ParseIP(loadbalancer),
+		},
+	}
+
+	cert, checksum, _, err := o.deployCertificate(ctx, certConfig, nil)
```

Comment:
the method `deployCertificate`. should be moved - it is currently in the file `etcd_certificates` but is used also to deploy API server certs.

Reply:
Additionally, the `deployCertificate`  should have a proper error message. here:
```
	cert, err := loadOrGenerateCertificateSecret(ctx, o.client, objectKey, certConfig)
	if err != nil {
		return nil, "", nil, fmt.Errorf("existing secret (%s/%s) contains an invalid x509 certificate. Please manually repair its contents as it should not be overwritten: %w", objectKey.Namespace, objectKey.Name, err)
	}
```

Adding it like this because github is not capable. of handling these amount of files in one diff.

Reply:
We have moved `deployCertificate` to `utils.go`.

We also added an error message which contains the secret namespace and name. But we have kept the message more general, because at this point there are also other possible reasons for the error. For example, reading or updating the secret might have failed; the certificate is not necessarily invalid.